loopy/target/pyopencl.py

```diff
@@ -244,18 +244,30 @@ def map_sum(self, expr, type_context):
 
             c_applied = [self.rec(c, type_context, tgt_dtype) for c in complexes]
 
+            mul_name = f"{tgt_name}_mul"
+
             def binary_tree_add(start, end):
                 if start + 1 == end:
                     return c_applied[start]
                 mid = (start + end)//2
                 lsum = binary_tree_add(start, mid)
                 rsum = binary_tree_add(mid, end)
-                return p.Variable("%s_add" % tgt_name)(lsum, rsum)
+
+                if isinstance(lsum, p.Call) and isinstance(lsum.function,
+                        p.Variable) and lsum.function.name == mul_name:
+                    return p.Variable(f"{tgt_name}_fma")(*lsum.parameters, rsum)
+
+                elif isinstance(rsum, p.Call) and isinstance(rsum.function,
+                        p.Variable) and rsum.function.name == mul_name:
+                    return p.Variable(f"{tgt_name}_fma")(*rsum.parameters, lsum)
+
+                else:
+                    return p.Variable(f"{tgt_name}_add")(lsum, rsum)
```

Comment:
@kaushikcfd Do you think `fma` generation should be an option? Or should we enable it by default?

Reply:
Given that the `_fma` variant never leads to a slow-down compared to its alternative, always emitting the FMA variant seems like a good choice.

Reply:
Nvm, saw the `allow_reordering` option in [9eaac97](https://github.com/inducer/loopy/pull/698/commits/9eaac973c524b674bcacf8674ee59732190255c5). Looks good.